src/vmm/src/vstate/vm.rs

```diff
@@ -360,40 +366,81 @@ impl Vm {
         Ok(())
     }
 
+    pub(crate) fn register_kvm_region(
+        &self,
+        region: kvm_userspace_memory_region,
+    ) -> Result<(), VmError> {
+        // SAFETY: Safe because the fd is a valid KVM file descriptor.
+        unsafe {
+            self.fd()
+                .set_user_memory_region(region)
+                .map_err(VmError::SetUserMemoryRegion)?;
+        }
+
+        Ok(())
+    }
+
     /// Register a new memory region to this [`Vm`].
     pub fn register_memory_region(&mut self, region: GuestRegionMmap) -> Result<(), VmError> {
-        let next_slot = self
-            .guest_memory()
-            .num_regions()
-            .try_into()
-            .expect("Number of existing memory regions exceeds u32::MAX");
-        if self.common.max_memslots <= next_slot {
+        let arcd_region = Arc::new(GuestMemoryRegionExt::dram_from_mmap_region(
+            region,
+            self.common.next_slot,
+        ));
+
+        self.common.next_slot += 1;
+        if self.common.next_slot > self.common.max_memslots {
             return Err(VmError::NotEnoughMemorySlots(self.common.max_memslots));
         }
 
-        let flags = if region.bitmap().is_some() {
-            KVM_MEM_LOG_DIRTY_PAGES
-        } else {
-            0
-        };
+        let new_guest_memory = self
+            .common
+            .guest_memory
+            .insert_region(Arc::clone(&arcd_region))?;
 
-        let memory_region = kvm_userspace_memory_region {
-            slot: next_slot,
-            guest_phys_addr: region.start_addr().raw_value(),
-            memory_size: region.len(),
-            userspace_addr: region.as_ptr() as u64,
-            flags,
-        };
+        self.register_kvm_region(arcd_region.kvm_userspace_memory_region())?;
 
-        let new_guest_memory = self.common.guest_memory.insert_region(Arc::new(region))?;
+        self.common.guest_memory = new_guest_memory;
 
-        // SAFETY: Safe because the fd is a valid KVM file descriptor.
-        unsafe {
-            self.fd()
-                .set_user_memory_region(memory_region)
-                .map_err(VmError::SetUserMemoryRegion)?;
+        Ok(())
+    }
+
+    /// Register a list of new memory regions to this [`Vm`].
+    pub fn restore_memory_regions(
+        &mut self,
+        regions: Vec<GuestRegionMmap>,
+        state: &GuestMemoryState,
```

Comment:
As I can see, the `state` is only used for `GuestRegionType` information. Can it be replaced by it? `state: &GuestMemoryState` -> `type: GuestRegionType`